### PR TITLE
windows builds

### LINKED ISSUE #1: Windows build
Note that shogun uses C++11 and doesn't compile 32-bit, so this is only going to be possible for win-64, python 3.5+.

Issues:

- [x] `bigobj` compile error: fixed in https://github.com/shogun-toolbox/shogun/commit/72cfeb15639d4b39565d61700dce11c0d839c4da, add as patch if for 6.0.0
- [x] `algorithm` header missing: fixed in https://github.com/shogun-toolbox/shogun/commit/8bf4aed8252e1c3772d249eb8b854de3286dc2e0, add as patch if for 6.0.0
- [ ] Not linking the examples correctly when there are optional shared libraries present: https://github.com/shogun-toolbox/shogun/issues/3917
- [ ] Even with the shared lib dependencies turned off, AppVeyor [ran out of space](https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.14895/job/j7175cu857satqxa); probably need to talk to them about that
- [ ] BLAS/LAPACK support: not a huge deal because it mostly uses Eigen
  - MKL option: depends on https://github.com/conda-forge/staged-recipes/issues/3022, and would probably also need to distribute the future [BSD version of shogun](https://github.com/shogun-toolbox/shogun/projects/7)
  - Another: OpenBLAS, depending on https://github.com/conda-forge/openblas-feedstock/issues/2

### PR DESCRIPTION
**Pull request by djsutherland**
Fixes #1, mostly.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -90,6 +90,9 @@ build:
     - msbuild shogun.sln                               # [win]
     - if errorlevel 1 exit 1                           # [win]
 
+    - cmake --build . --target install -config Release  # [win]
```
Comment (vigsterkr) [inline recipe/meta.yaml]: it's `--config` with double-dash

File: recipe/meta.yaml [hunk 2]
```diff
@@ -26,72 +27,109 @@ source:
...
+          -DCMAKE_INSTALL_PREFIX="%LIBRARY_PREFIX%" ^  # [win]
+          -DCMAKE_BUILD_TYPE=Release ^                 # [win]
+          -DBUILD_EXAMPLES=OFF ^                       # [win]
+          -DBUILD_META_EXAMPLES=OFF ^                  # [win]
+          -DBUNDLE_JSON=OFF ^                          # [win]
+          -DBUNDLE_NLOPT=OFF ^                         # [win]
+          -DUSE_SVMLIGHT=OFF ^                         # [win]
+          -DENABLE_TESTING=OFF ^                       # [win]
+          -DENABLE_COVERAGE=OFF ^                      # [win]
+          -DLICENSE_GPL_SHOGUN=ON ^                    # [win]
```
Comment (vigsterkr) [inline recipe/meta.yaml]: not necessary anymore with 6.1.1

File: recipe/meta.yaml [hunk 3]
```diff
@@ -26,72 +27,109 @@ source:
...
+          -DLIBSHOGUN_BUILD_STATIC=ON                  # [win]
```
Comment (vigsterkr) [inline recipe/meta.yaml]: not necessary anymore with 6.1.1

File: recipe/meta.yaml [hunk 4]
```diff
@@ -26,72 +27,109 @@ source:
...
+    - msbuild src\\shogun\\shogun.vcxproj              # [win]
+    - if errorlevel 1 exit 1                           # [win]
+
+    - msbuild src\\shogun\\shogun-static.vcxproj       # [win]
+    - if errorlevel 1 exit 1                           # [win]
+
+    - msbuild shogun.sln                               # [win]
+    - if errorlevel 1 exit 1                           # [win]
+
+    - cmake --build . --target install -config Release  # [win]
```
Comment (vigsterkr) [inline recipe/meta.yaml]: `--config`

**Review by vigsterkr — CHANGES_REQUESTED** [review]

File: recipe/meta.yaml [hunk 5]
```diff
@@ -26,72 +27,109 @@ source:
...
+          -DBUNDLE_JSON=OFF ^                          # [win]
+          -DBUNDLE_NLOPT=OFF ^                         # [win]
+          -DUSE_SVMLIGHT=OFF ^                         # [win]
+          -DENABLE_TESTING=OFF ^                       # [win]
+          -DENABLE_COVERAGE=OFF ^                      # [win]
+          -DLICENSE_GPL_SHOGUN=ON ^                    # [win]
+          -DLIBSHOGUN_BUILD_STATIC=ON                  # [win]
+    - if errorlevel 1 exit 1                           # [win]
+
+    - msbuild src\\shogun\\libshogun.vcxproj           # [win]
```
Comment (vigsterkr) [inline recipe/meta.yaml]: add the configuration `Release`:
```
msbuild /p:Configuration=Release src\\shogun\\libshogun.vcxproj
```

File: recipe/meta.yaml [hunk 6]
```diff
@@ -26,72 +27,109 @@ source:
...
+    - if errorlevel 1 exit 1                           # [win]
+
+    - msbuild src\\shogun\\shogun.vcxproj              # [win]
```
Comment (vigsterkr) [inline recipe/meta.yaml]: ```msbuild /p:Configuration=Release src\\shogun\\shogun.vcxproj```

File: recipe/meta.yaml [hunk 7]
```diff
@@ -26,72 +27,109 @@ source:
...
+    - if errorlevel 1 exit 1                           # [win]
+
+    - msbuild src\\shogun\\shogun-static.vcxproj       # [win]
```
Comment (vigsterkr) [inline recipe/meta.yaml]:
```
msbuild /p:Configuration=Release src\\shogun\\shogun-static.vcxproj
```

File: recipe/meta.yaml [hunk 8]
```diff
@@ -26,72 +27,109 @@ source:
...
+    - if errorlevel 1 exit 1                           # [win]
+
+    - msbuild shogun.sln                               # [win]
```
Comment (vigsterkr) [inline recipe/meta.yaml]:
```
msbuild /p:Configuration=Release shogun.sln
```

**Comment by vigsterkr** [comment]
ok the error is because of the `minimal`: https://ci.appveyor.com/project/conda-forge/shogun-cpp-feedstock/build/1.0.13#L16958
```
  CMake Error at examples/minimal/cmake_install.cmake:36 (file):
    file INSTALL cannot find
    "C:/bld/shogun-cpp_1511924007359/work/shogun-shogun_6.1.1/build/examples/minimal/minimal".
  Call Stack (most recent call first):
    cmake_install.cmake:38 (include)
```

**Comment by vigsterkr** [comment]
this should fix the issue:
https://github.com/shogun-toolbox/shogun/commit/93a64bdca066aa8320f0f468536014422a614b00

**Comment by vigsterkr** [comment]
ok i've created 6.1.2 release as i've messed up the commits in this early morning a bit :)

**Comment by djsutherland** [comment]
@vigsterkr Okay, trying again....

btw the Python interface of 6.1.1 is up for Linux and should be up for Mac in a few minutes.

**Comment by vigsterkr** [comment]
@dougalsutherland hehe yueah i know i've been using the linux version for the last couple of hours ;) thnx heaps!

**Comment by vigsterkr** [comment]
@dougalsutherland the test is not good for windows:
```
(C:\bld\shogun-cpp_1511949224412\_t_env) (root) C:\bld\shogun-cpp_1511949224412\test_tmp>python -c "import os, ctypes; ctypes.cdll[os.path.join(os.environ['LIBRARY_LIB'], 'libshogun.dylib')]"
```

this definitely will not work - no matter what - on win32 :)
do you need such test when you just build the libshogun itself (that's not yet the python interface). anyhow i dont see the shogun.dll being installed so that might be a problem in the future :)

**Comment by djsutherland** [comment]
I screwed up the test at the very end, but seems to have worked until then, so going ahead and merging.

leap of faith

cc @vigsterkr :)

**Comment by vigsterkr** [comment]
@dougalsutherland hihi :)

**Comment by djsutherland** [comment]
Have a Windows PR ready to go for shogun-feedstock too, just waiting on shogun-cpp to build.

**Pull request merged**